bpfman-operator/apis/v1alpha1/shared_types.go

```diff
@@ -0,0 +1,383 @@
+	ByteCode ByteCodeSelector `json:"byteCode"`
+
+	// TODO: need to work out how MapOwnerSelector will work after load-attach-split
+	// mapOwnerSelector is used to select the loaded eBPF program this eBPF program
+	// will share a map with.
+	// +optional
+	MapOwnerSelector *metav1.LabelSelector `json:"mapOwnerSelector"`
+}
+
+// BpfAppStatus reflects the status of a BpfApplication or BpfApplicationState object
+type BpfAppStatus struct {
+	// For a BpfApplication object, Conditions contains the global cluster state
+	// for the object. For a BpfApplicationState object, Conditions contains the
+	// state of the BpfApplication object on the given node.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+}
+
+// AttachInfoStateCommon reflects the status for one attach point for a given bpf
+// application program
+type AttachInfoStateCommon struct {
+	// shouldAttach reflects whether the attachment should exist.
+	ShouldAttach bool `json:"shouldAttach"`
+	// uuid is an Unique identifier for the attach point assigned by bpfman agent.
+	UUID string `json:"uuid"`
+	// linkId is an identifier for the link assigned by bpfman. This field is
+	// empty until the program is successfully attached and bpfman returns the
+	// id.
+	LinkId *uint32 `json:"linkId"`
+	// linkStatus reflects whether the attachment has been reconciled
+	// successfully, and if not, why.
+	LinkStatus LinkStatus `json:"linkStatus"`
+}
+
+type BpfProgramStateCommon struct {
+	// name is the name of the function that is the entry point for the BPF
+	// program
+	Name string `json:"name"`
+	// programLinkStatus reflects whether all links requested for the program
+	// are in the correct state.
+	ProgramLinkStatus ProgramLinkStatus `json:"programLinkStatus"`
+	// programId is the id of the program in the kernel.  Not set until the
+	// program is loaded.
+	// +optional
+	ProgramId *uint32 `json:"programId"`
+}
+
+// PullPolicy describes a policy for if/when to pull a container image
+// +kubebuilder:validation:Enum=Always;Never;IfNotPresent
+type PullPolicy string
+
+const (
+	// PullAlways means that bpfman always attempts to pull the latest bytecode image. Container will fail If the pull fails.
+	PullAlways PullPolicy = "Always"
+	// PullNever means that bpfman never pulls an image, but only uses a local image. Container will fail if the image isn't present
+	PullNever PullPolicy = "Never"
+	// PullIfNotPresent means that bpfman pulls if the image isn't present on disk. Container will fail if the image isn't present and the pull fails.
+	PullIfNotPresent PullPolicy = "IfNotPresent"
+)
+
+// ByteCodeSelector defines the various ways to reference bpf bytecode objects.
+type ByteCodeSelector struct {
+	// image used to specify a bytecode container image.
+	Image *ByteCodeImage `json:"image,omitempty"`
```

Comment:
The field definition here maps to the conventions for an optional field, but it is overridden by the package level required. Should this be optional or required?

Reply:
it can be specified via image or path ?

Reply:
So you want an option to use image _or_ path? If that is the case, that is not what your API currently states. It states that you must set _both_ image _and_ path.

Reply:
missing optional ? it doesn't complain atm when I pass just the Image though ?

Reply:
You have a package level required set so I'm surprised that is working as optional. Generally when you want to do a "one or the other" option you would use a discriminated union. Some docs/guidelines on a discriminated union: https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions

Reply:
Also, remove the package level required.

Reply:
ok